rfcs/20210115-tfx-periodic-training.md

```diff
@@ -0,0 +1,445 @@
+*   `root/day-2/attempt1/data` maps to Span 2 Version 1
+
+Note that the new generated data Version must have a bigger Version id. On the
+other hand, Span id semantics are up to the user: it can be the generation order
+(new generated data has bigger id), or the timeline order depends on the data
+content instead of generation time (e.g., each Span represents the data of a
+certain date). In the latter case, the Span of day 1 might be generated after
+the Span of day 2 due to the late arriving of day 1’s raw input data. The latest
+Span/Version below refers to the one with the highest id instead of the new
+generated one.
+
+For the purposes of this RFC (which is restricted to synchronous pipeline
+execution), ExampleGen generates one Span at a time. Normally, ExampleGen picks
+up the input data corresponding to the latest Span. However, for this RFC the
+recommended configuration for ExampleGen uses an explicitly specified Span id,
+which can (for example) be generated automatically from the timestamp of the
+periodic run or selected manually for a backfill run. Within a Span, ExampleGen
+always picks up the input data corresponding to the latest Version.
+
+Span and Version ids don’t need to be contiguous. For example, you can use only
+odd numbers. But for simplicity the following discussion assumes the ids are
+contiguous starting from 1.
+
+The following sections focus on the abstract Span concept, but we provide some
+examples with concrete Span semantics for better understanding.
+
+### Periodic Training Pipeline
+
+In this section, we discuss how a TFX periodic batch pipeline handles Spans and
+Versions. The scenarios we want to support are:
+
+*   Users can update the model to capture newly arrived input data.
+*   Users can update the model when previously processed input data is updated.
+*   Models can be trained based on the single Span generated in the current
+    pipeline run, or based on a rolling window of Spans including Spans
+    generated in previous pipeline runs.
+
+The design focuses on the TFX batch pipeline itself; it’s up to users to define
+how and when to schedule the batch pipeline for the updated input data.
+
+#### Single Span
+
+If model training only requires a single Span (rolling window size equals to 1),
+the setup is simple: TFX pipeline run should be periodically scheduled at the
+time that the new input data block is available. ExampleGen will pick up the new
+data, and the model will be trained based on the ExampleGen’s output Span.
+
+<div style="text-align:center"><img src='20210115-tfx-periodic-training/single_span.png', width='600'></div>
+
+*For example, assume a Span corresponds to a week of data, and every Monday, the
+previous week’s data is finalized. Users can set up a routine TFX pipeline to
+run every Wednesday, so that ExampleGen will pick up the previous week’s data
+and generate a corresponding output Span for training.*
+
+#### Rolling Window
```

Comment:
Not sure, but is the plan to support different windows per split in the dataset.

If I'm doing serialized data, I may want to train on N-5 to N-1 days and evaluate on N day.

Reply:
it depends on how you set up your pipeline, 
If Evaluator takes Resolver's output (span n to n-windowsize), it will use eval based on n spans,
if Evaluator takes ExampleGen's output (only have current span n), it will use span n to eval.

Reply:
When I meant evaluate I mean eval split. Aka splits in the dataset have different spans.

Reply:
each run of ExampleGen generates a span with multiple splits, e.g., train and eval.

Thus depends on how many span you get, if Evaluator takes span 1,2,3, it will get a train split (that contains span1train, span2train, span3train) and a eval split (contains span1eval, span2eval, span3eval)

Reply:
My use case was more of:

I have 5 spans {1,2,3,4,5}
- I want to use spans 1-4 for train split
- I want to use span 5 for eval split. 

Currently this is impossible due to range_config being a component level config instead of Split config.

Reply:
currently both sync/async follows the same [span/version/split concept](https://github.com/tensorflow/tfx/blob/master/docs/guide/examplegen.md#span-version-and-split)

so if you do want to have different split config, you need to have 2 examplegens, this is same for sync and async.

Note that for model training, if it take span 1-4, the training and testing dataset must from those span 1-4
but for Evaluator, it can take a different dataset compared to training, and it can configure which split to use in that dataset.

So if you can have the same splits for all spans, you can setup the sync pipeline like this,
span 1-5 has train/eval/test, span 1-5's train/eval is provided to training to train a model, span 5's test is provided to evaluator for evaluation